File: c/virtual_stack.c
/* 仮想スタック機械への翻訳。課題 1〜3 は not_implemented を埋めること。 */

#include "virtual_stack.h"
#include "util.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

void code_push(Code *c, Instr ins) {
    if (c->n >= c->cap) {
        c->cap = c->cap ? c->cap * 2 : 8;
        c->data = realloc(c->data, (size_t)c->cap * sizeof(Instr));
        if (!c->data)
            die("out of memory");
    }
    c->data[c->n++] = ins;
}

void code_concat(Code *dst, Code src) {
    for (int i = 0; i < src.n; i++)
        code_push(dst, src.data[i]);
}

Instr instr_push(int n) { return (Instr){.kind = I_PUSH, .n = n}; }
Instr instr_rvalue(const char *name) {
    return (Instr){.kind = I_RVALUE, .name = (char *)name};
}
Instr instr_lpush(const char *name) {
    return (Instr){.kind = I_LPUSH, .name = (char *)name};
}
Instr instr_op(InstrKind k) { return (Instr){.kind = k}; }
Instr instr_label_test(const char *test, const char *out) {
    return (Instr){.kind = I_LABEL_TEST, .test = (char *)test, .out = (char *)out};
}
Instr instr_label_out(const char *test, const char *out) {
    return (Instr){.kind = I_LABEL_OUT, .test = (char *)test, .out = (char *)out};
}
Instr instr_goto(const char *label) {
    return (Instr){.kind = I_GOTO, .name = (char *)label};
}
Instr instr_gofalse(const char *label) {
    return (Instr){.kind = I_GOFALSE, .name = (char *)label};
}

static int label_count = -1;

char *gen_label(void) {
    char buf[32];
    label_count++;
    snprintf(buf, sizeof buf, "L.%d", label_count);
    return xstrdup(buf);
}

void reset_labels(void) { label_count = -1; }

static Code bin_arith(Arith *a, Instr op) {
    Code c = compile_arith(a->lhs);
    code_concat(&c, compile_arith(a->rhs));
    code_push(&c, op);
    return c;
}

static Code bin_pred_arith(Pred *p, Instr op) {
    Code c = compile_arith(p->a);
    code_concat(&c, compile_arith(p->b));
    code_push(&c, op);
    return c;
}

Code compile_arith(Arith *a) {
    Code c = {0};
    switch (a->kind) {
    case ARITH_VAR:
        code_push(&c, instr_rvalue(a->name));
        return c;
    case ARITH_NUM:
        code_push(&c, instr_push(a->n));
        return c;
    case ARITH_ADD:
        return bin_arith(a, instr_op(I_PLUS));
    case ARITH_SUB:
        not_implemented(
            "課題1: Sub を仮想スタック命令へ翻訳してください。"
            "Add の場合を参考に、最後に I_MINUS を置きます。");
        return c;
    case ARITH_MUL:
        not_implemented(
            "課題1: Mul を仮想スタック命令へ翻訳してください。"
            "Add の場合を参考に、最後に I_TIMES を置きます。");
        return c;
    case ARITH_DIV:
        not_implemented(
            "課題1: Div を仮想スタック命令へ翻訳してください。"
            "Add の場合を参考に、最後に I_DIVIDE を置きます。");
        return c;
    }
    return c;
}

Code compile_predicate(Pred *p) {
    Code c = {0};
    switch (p->kind) {
    case PRED_TRUE:
        code_push(&c, instr_op(I_TRUE));
        return c;
    case PRED_FALSE:
        code_push(&c, instr_op(I_FALSE));
        return c;
    case PRED_NOT:
        c = compile_predicate(p->lhs);
        code_push(&c, instr_op(I_NOT));
        return c;
    case PRED_AND:
        c = compile_predicate(p->lhs);
        code_concat(&c, compile_predicate(p->rhs));
        code_push(&c, instr_op(I_AND));
        return c;
    case PRED_OR:
        c = compile_predicate(p->lhs);
        code_concat(&c, compile_predicate(p->rhs));
        code_push(&c, instr_op(I_OR));
        return c;
    case PRED_LT:
        return bin_pred_arith(p, instr_op(I_LT));
    case PRED_LE:
        not_implemented(
            "課題2: LE を仮想スタック命令へ翻訳してください。"
            "LT の場合を参考に、最後に I_LE を置きます。");
        return c;
    case PRED_GT:
        not_implemented(
            "課題2: GT を仮想スタック命令へ翻訳してください。"
            "LT の場合を参考に、最後に I_GT を置きます。");
        return c;
    case PRED_GE:
        not_implemented(
            "課題2: GE を仮想スタック命令へ翻訳してください。"
            "LT の場合を参考に、最後に I_GE を置きます。");
        return c;
    case PRED_EQ:
        not_implemented(
            "課題2: EQ を仮想スタック命令へ翻訳してください。"
            "LT の場合を参考に、最後に I_EQ を置きます。");
        return c;
    }
    return c;
}

Code compile_statement(Stmt *s) {
    Code c = {0};
    switch (s->kind) {
    case STMT_ASSIGN:
        c = compile_arith(s->arith);
        code_push(&c, instr_lpush(s->name));
        return c;
    case STMT_SKIP:
        return c;
    case STMT_PRINT:
        c = s->print_is_pred ? compile_predicate(s->pred) : compile_arith(s->arith);
        code_push(&c, instr_op(I_PRINT));
        return c;
    case STMT_BLOCK:
        return compile_statement(s->first);
    case STMT_SEQ:
        c = compile_statement(s->first);
        code_concat(&c, compile_statement(s->second));
        return c;
    case STMT_WHILE:
        /* char *test = gen_label();
         * char *out = gen_label();
         * Code p = compile_predicate(s->pred);
         * Code body = compile_statement(s->second);
         * 上で作った LabelTest, p, GoFalse, body, GoTo, LabelOut を
         * どの順で並べると while になるか、README の翻訳例を見て書いてください。
         */
        not_implemented(
            "課題3: while をラベルとジャンプへ翻訳してください。\n"
            "README の while 翻訳例と、すぐ上のコメントを見てください。");
        return c;
    case STMT_IF:
        not_implemented(
            "発展課題: If を実装してください。"
            "条件のあとに I_IF_START, then 節, I_ELSE, else 節, I_IF_END です。");
        return c;
    }
    return c;
}

Code compile_stack(Stmt *s) {
    reset_labels();
    return compile_statement(s);
}

char *format_instr(Instr ins) {
    Buf b;
    buf_init(&b);
    switch (ins.kind) {
    case I_LPUSH: buf_printf(&b, "lpush\t%s", ins.name); break;
    case I_RVALUE: buf_printf(&b, "rvalue\t%s", ins.name); break;
    case I_PUSH: buf_printf(&b, "push\t%d", ins.n); break;
    case I_PLUS: buf_printf(&b, "+"); break;
    case I_MINUS: buf_printf(&b, "-"); break;
    case I_TIMES: buf_printf(&b, "*"); break;
    case I_DIVIDE: buf_printf(&b, "/"); break;
    case I_LT: buf_printf(&b, "<"); break;
    case I_LE: buf_printf(&b, "<="); break;
    case I_GT: buf_printf(&b, ">"); break;
    case I_GE: buf_printf(&b, ">="); break;
    case I_EQ: buf_printf(&b, "=="); break;
    case I_NOT: buf_printf(&b, "not"); break;
    case I_AND: buf_printf(&b, "and"); break;
    case I_OR: buf_printf(&b, "or"); break;
    case I_TRUE: buf_printf(&b, "true"); break;
    case I_FALSE: buf_printf(&b, "false"); break;
    case I_LABEL_TEST: buf_printf(&b, "label\t%s", ins.test); break;
    case I_LABEL_OUT: buf_printf(&b, "label\t%s", ins.out); break;
    case I_GOFALSE: buf_printf(&b, "gofalse\t%s", ins.name); break;
    case I_GOTO: buf_printf(&b, "goto\t%s", ins.name); break;
    case I_PRINT: buf_printf(&b, "print"); break;
    case I_IF_START: buf_printf(&b, "if"); break;
    case I_ELSE: buf_printf(&b, "else"); break;
    case I_IF_END: buf_printf(&b, "endif"); break;
    }
    return buf_take(&b);
}

char *format_code(Code code) {
    Buf b;
    buf_init(&b);
    for (int i = 0; i < code.n; i++)
        buf_printf(&b, "%s\n", format_instr(code.data[i]));
    return buf_take(&b);
}

File: python/virtual_stack.py
"""WhileLang の AST を、スタック機械の命令列へ変換する。

課題 1〜3 はここの NotImplementedError を埋めること。
"""

from __future__ import annotations

from dataclasses import dataclass

from syntax import (
    Add,
    And,
    Arith,
    Assign,
    Block,
    Div,
    EQ,
    FalseConst,
    GE,
    GT,
    If,
    LE,
    LT,
    Mul,
    Not,
    Num,
    Or,
    Pred,
    Print,
    Seq,
    Skip,
    Stmt,
    Sub,
    TrueConst,
    Var,
    While,
)


# --- 命令 -------------------------------------------------------------------

@dataclass(frozen=True)
class Push:
    n: int


@dataclass(frozen=True)
class RValue:
    name: str


@dataclass(frozen=True)
class LPush:
    name: str


@dataclass(frozen=True)
class Plus:
    pass


@dataclass(frozen=True)
class Minus:
    pass


@dataclass(frozen=True)
class Times:
    pass


@dataclass(frozen=True)
class Divide:
    pass


@dataclass(frozen=True)
class Eq:
    pass


@dataclass(frozen=True)
class Lt:
    pass


@dataclass(frozen=True)
class Le:
    pass


@dataclass(frozen=True)
class Gt:
    pass


@dataclass(frozen=True)
class Ge:
    pass


@dataclass(frozen=True)
class NotOp:
    pass


@dataclass(frozen=True)
class AndOp:
    pass


@dataclass(frozen=True)
class OrOp:
    pass


@dataclass(frozen=True)
class TrueOp:
    pass


@dataclass(frozen=True)
class FalseOp:
    pass


@dataclass(frozen=True)
class PrintOp:
    pass


@dataclass(frozen=True)
class LabelTest:
    test: str
    out: str


@dataclass(frozen=True)
class LabelOut:
    test: str
    out: str


@dataclass(frozen=True)
class GoTo:
    label: str


@dataclass(frozen=True)
class GoFalse:
    label: str


@dataclass(frozen=True)
class IfStart:
    pass


@dataclass(frozen=True)
class ElseOp:
    pass


@dataclass(frozen=True)
class IfEnd:
    pass


Instr = (
    Push
    | RValue
    | LPush
    | Plus
    | Minus
    | Times
    | Divide
    | Eq
    | Lt
    | Le
    | Gt
    | Ge
    | NotOp
    | AndOp
    | OrOp
    | TrueOp
    | FalseOp
    | PrintOp
    | LabelTest
    | LabelOut
    | GoTo
    | GoFalse
    | IfStart
    | ElseOp
    | IfEnd
)


_count = -1


def gen_label() -> str:
    global _count
    _count += 1
    return f"L.{_count}"


def reset() -> None:
    global _count
    _count = -1


def compile_arith(arith: Arith) -> list[Instr]:
    match arith:
        case Var(name):
            return [RValue(name)]
        case Num(n):
            return [Push(n)]
        case Add(lhs, rhs):
            return compile_arith(lhs) + compile_arith(rhs) + [Plus()]
        case Sub(lhs, rhs):
            raise NotImplementedError(
                "課題1: Sub を仮想スタック命令へ翻訳してください。"
                "Add の場合を参考に、最後に Minus() を置きます。"
            )
        case Mul(lhs, rhs):
            raise NotImplementedError(
                "課題1: Mul を仮想スタック命令へ翻訳してください。"
                "Add の場合を参考に、最後に Times() を置きます。"
            )
        case Div(lhs, rhs):
            raise NotImplementedError(
                "課題1: Div を仮想スタック命令へ翻訳してください。"
                "Add の場合を参考に、最後に Divide() を置きます。"
            )


def compile_predicate(pred: Pred) -> list[Instr]:
    match pred:
        case TrueConst():
            return [TrueOp()]
        case FalseConst():
            return [FalseOp()]
        case Not(p):
            return compile_predicate(p) + [NotOp()]
        case And(lhs, rhs):
            return compile_predicate(lhs) + compile_predicate(rhs) + [AndOp()]
        case Or(lhs, rhs):
            return compile_predicate(lhs) + compile_predicate(rhs) + [OrOp()]
        case LT(lhs, rhs):
            return compile_arith(lhs) + compile_arith(rhs) + [Lt()]
        case LE(lhs, rhs):
            raise NotImplementedError(
                "課題2: LE を仮想スタック命令へ翻訳してください。"
                "LT の場合を参考に、最後に Le() を置きます。"
            )
        case GT(lhs, rhs):
            raise NotImplementedError(
                "課題2: GT を仮想スタック命令へ翻訳してください。"
                "LT の場合を参考に、最後に Gt() を置きます。"
            )
        case GE(lhs, rhs):
            raise NotImplementedError(
                "課題2: GE を仮想スタック命令へ翻訳してください。"
                "LT の場合を参考に、最後に Ge() を置きます。"
            )
        case EQ(lhs, rhs):
            raise NotImplementedError(
                "課題2: EQ を仮想スタック命令へ翻訳してください。"
                "LT の場合を参考に、最後に Eq() を置きます。"
            )


def compile_statement(stmt: Stmt) -> list[Instr]:
    match stmt:
        case Assign(name, expr):
            return compile_arith(expr) + [LPush(name)]
        case Skip():
            return []
        case Print(expr):
            if isinstance(expr, (TrueConst, FalseConst, Not, And, Or, LT, LE, GT, GE, EQ)):
                return compile_predicate(expr) + [PrintOp()]
            return compile_arith(expr) + [PrintOp()]
        case Block(body):
            return compile_statement(body)
        case Seq(first, second):
            return compile_statement(first) + compile_statement(second)
        case While(pred, body):
            # test = gen_label()
            # out = gen_label()
            # p = compile_predicate(pred)
            # s = compile_statement(body)
            # 上で作った LabelTest, p, GoFalse, s, GoTo, LabelOut を
            # どの順で並べると while になるか、README の翻訳例を見て書いてください。
            raise NotImplementedError(
                "課題3: while をラベルとジャンプへ翻訳してください。\n"
                "README の while 翻訳例と、すぐ上のコメントを見てください。"
            )
        case If(_pred, _then, _else):
            raise NotImplementedError(
                "発展課題: If を実装してください。"
                "条件のあとに IfStart(), then 節, ElseOp(), else 節, IfEnd() です。"
            )


def compile_stack(stmt: Stmt) -> list[Instr]:
    reset()
    return compile_statement(stmt)


def format_instr(ins: Instr) -> str:
    match ins:
        case LPush(name):
            return f"lpush\t{name}"
        case RValue(name):
            return f"rvalue\t{name}"
        case Push(n):
            return f"push\t{n}"
        case Plus():
            return "+"
        case Minus():
            return "-"
        case Times():
            return "*"
        case Divide():
            return "/"
        case Lt():
            return "<"
        case Le():
            return "<="
        case Gt():
            return ">"
        case Ge():
            return ">="
        case Eq():
            return "=="
        case NotOp():
            return "not"
        case AndOp():
            return "and"
        case OrOp():
            return "or"
        case TrueOp():
            return "true"
        case FalseOp():
            return "false"
        case LabelTest(test, _out):
            return f"label\t{test}"
        case LabelOut(_test, out):
            return f"label\t{out}"
        case GoFalse(label):
            return f"gofalse\t{label}"
        case GoTo(label):
            return f"goto\t{label}"
        case PrintOp():
            return "print"
        case IfStart():
            return "if"
        case ElseOp():
            return "else"
        case IfEnd():
            return "endif"


def format_code(code: list[Instr]) -> str:
    return "".join(format_instr(ins) + "\n" for ins in code)
